flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/architecture/ApiRules.java

```diff
@@ -0,0 +1,93 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.testutils.architecture;
+
+import org.apache.flink.annotation.Public;
+import org.apache.flink.annotation.PublicEvolving;
+
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchRule;
+
+import static com.tngtech.archunit.core.domain.JavaClass.Predicates.resideOutsideOfPackage;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.methods;
+import static org.apache.flink.testutils.architecture.Predicates.areAnnotatedWithAtLeastOneOf;
+import static org.apache.flink.testutils.architecture.Predicates.haveLeafArgumentTypes;
+import static org.apache.flink.testutils.architecture.Predicates.haveLeafReturnTypes;
+
+/** Architecture rules for APIs. */
+public class ApiRules {
+
+    /**
+     * Ensure that public methods marked {@link Public} only use return and argument types which are
+     * also marked {@link Public}.
+     *
+     * <p>Note that this is a stronger requirement than {@link
+     * #PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES} and ensures that {@link
+     * Public} APIs do not use {@link PublicEvolving} APIs.
+     */
+    @ArchTest
+    public static final ArchRule PUBLIC_API_METHODS_USE_ONLY_PUBLIC_API_TYPES =
+            methods()
+                    .that()
+                    .areDeclaredInClassesThat()
+                    .areAnnotatedWith(Public.class)
+                    .and()
+                    .arePublic()
+                    .should(
+                            haveLeafReturnTypes(
+                                    resideOutsideOfPackage("org.apache.flink..")
+                                            .or(
+                                                    areAnnotatedWithAtLeastOneOf(
+                                                            Public.class, Deprecated.class))))
+                    .andShould(
+                            haveLeafArgumentTypes(
+                                    resideOutsideOfPackage("org.apache.flink..")
+                                            .or(
+                                                    areAnnotatedWithAtLeastOneOf(
+                                                            Public.class, Deprecated.class))));
+
+    /**
+     * Ensure that public methods marked {@link Public} or {@link PublicEvolving} only use return
+     * and argument types which are also marked {@link Public} or {@link PublicEvolving}.
+     */
+    @ArchTest
+    public static final ArchRule PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES =
```

Comment:
In flink-table-common this test currently identifies quite a few problems, e.g. a lot of methods in `Catalog` referencing non-annotated classes, or `SinkProvider` being PublicEvolving, but `Sink` being Experimental.